### PR TITLE
Consume bootstrap data to add applications to machines

### PR DESCRIPTION
**Pull request by Cynerva**
Fetch the bootstrap data, and for each application listed in the bootstrap data, check to see if a unit already exists on the machine. If not, add a new unit to the machine.

Testing this with the sample manifests in the control-plane-provider repo, I end up with the following model status:

```
Model               Controller                    Cloud/Region               Version  SLA          Timestamp
jujucluster-sample  jujucluster-sample-k8s-cloud  jujucluster-sample/Boston  3.0.2    unsupported  15:22:33-06:00

App                       Version  Status   Scale  Charm                     Channel      Rev  Exposed  Message
easyrsa                   3.0.1    active       1  easyrsa                   1.26/stable   33  no       Certificate Authority connected.
etcd                      3.4.22   active       1  etcd                      1.26/stable  726  no       Healthy with 1 known peer
kubeapi-load-balancer     1.18.0   active       1  kubeapi-load-balancer     1.26/stable   56  no       Loadbalancer ready.
kubernetes-control-plane  1.26.2   waiting      1  kubernetes-control-plane  1.26/stable  240  no       Waiting for workers.
kubernetes-worker                  unknown      0  kubernetes-worker         1.26/stable   92  no

Unit                         Workload  Agent  Machine  Public address  Ports         Message
easyrsa/0*                   active    idle   0        10.246.154.224                Certificate Authority connected.
etcd/0*                      active    idle   2        10.246.154.102  2379/tcp      Healthy with 1 known peer
kubeapi-load-balancer/0*     active    idle   1        10.246.154.113  443,6443/tcp  Loadbalancer ready.
kubernetes-control-plane/0*  waiting   idle   2        10.246.154.102                Waiting for workers.

Machine  State    Address         Inst id        Base          AZ  Message
0        started  10.246.154.224  juju-c2ee59-0  ubuntu@22.04      poweredOn
1        started  10.246.154.113  juju-c2ee59-1  ubuntu@22.04      poweredOn
2        started  10.246.154.102  juju-c2ee59-2  ubuntu@22.04      poweredOn
```

We still need to update JujuMachine.Status, but I wasn't quite sure how to approach that given that kubernetes-control-plane is stuck in waiting status without workers. Perhaps we should rework kubernetes-control-plane so that it can enter "active" status without workers?

### CODE REVIEW

**Comment by stonepreston** [comment]
This looks great. Regarding the setting of the ready status on the machine after the add-unit process is complete, the control plane provider will need to be able to 1. get the kubeconfig via the get-kubeconfig action from the control plane leader unit and 2. contact the api server via said kubeconfig. If both those things are true with things in this waiting state, then I think this would still be okay. 

That being said, I suppose to does make sense for the control plane charm to go active on its own without having a worker present, but not sure how many worms are in that can.

**Review by stonepreston — CHANGES_REQUESTED** [review]
Looks great, but one issue. Deletion logic will need to be updated to remove any present units before destroying the machine. 

Currently when the machine tries to be deleted the juju api returns an error:
```
"error": "machine 2 has unit \"kubernetes-control-plane/0\" assigned
```

We could probably use force and get around this, but probably cleaner to just loop over the existing units and remove em

**Comment by Cynerva** [comment]
I pushed some quick fixes to get things moving along. With charms from 1.27/edge, the Kubernetes API comes up, but the charm still gets stuck in waiting status.

I'll circle back and clean up the DestroyMachine call after we get other work unblocked.

I've opened https://github.com/charmed-kubernetes/charm-kubernetes-control-plane/pull/272 to remove the "Waiting for workers." status. The charm still won't enter Active status, though, because of the ["Missing CNI relation or config" status](https://github.com/charmed-kubernetes/charm-kubernetes-control-plane/blob/c45517faacea70a314cf93dc9ac0c67ef0407e7f/reactive/kubernetes_control_plane.py#L902). We probably need a config option to disable that check.

**Comment by stonepreston** [comment]
PR that adds config option to gnore missing cni config and enter active status without one is [here](https://github.com/charmed-kubernetes/charm-kubernetes-control-plane/pull/275)

File: controllers/jujucluster_controller.go [hunk 1]
```diff
@@ -865,23 +873,31 @@ func createApplicationsIfNeeded(ctx context.Context, jujuCluster *infrastructure
 			ApplicationName: "kubernetes-control-plane",
 			ModelUUID:       modelUUID,
 			CharmName:       "kubernetes-control-plane",
-			CharmChannel:    "1.26/stable",
+			CharmChannel:    "1.27/edge",
 			CharmBase:       "ubuntu@22.04",
 			Units:           0,
```
Comment (stonepreston) [inline controllers/jujucluster_controller.go:878]:
```suggestion
			Units:           0,
			Config: map[string]interface{}{
				"ignore-missing-cni": true,
				"enable-metrics":     false,
				"dns-provider":       "none",
			},
```

Comment (stonepreston) [reply controllers/jujucluster_controller.go:878]: kcp charm has been updated to ignore missing CNI, together with your changes for ignoring missing workers, the charm will now reach active/idle once units are deployed. 

note that metrics and dns must be disabled as well, as these pods cannot start without a CNI.

Comment (stonepreston) [reply controllers/jujucluster_controller.go:878]: now that kcp is going active/idle, we could wait for this condition in the machine controller and once its met, set ready status.

**Comment by Cynerva** [comment]
I did a quick test with the latest changes, and the JujuMachine is now reaching Ready status after the Juju applications reach active/idle status.

**Review by stonepreston — APPROVED** [review]

**Pull request merged**